Add new field Room

kangzongxian opened the pull request:
This denotes the room that the Guest is staying at. This is because the room is a quintessential element for a hotel.
closes #92 

Edits Made:

Prefix rm/
Add field "Room" to the Guest Class
Room field accepts hyphens and alphanumeric characters only. However, starting with a hyphen is invalid.
Allow edit of Room field
Add test cases to test the Room field

kangzongxian committed:
Added new field Room

This denotes the room that the Guest is staying at.
This is because the room is a quintessential element for a hotel.

bryanhce commented on docs/UserGuide.md:
```diff
@@ -192,12 +192,12 @@ _Details coming soon ..._
+| **Add**    | `add n/NAME p/PHONE_NUMBER e/EMAIL rm/ROOM dr/DATE_RANGE ng/NUMBER_OF_GUESTS rc/IS_ROOM_CLEAN ​` <br> e.g., `add n/John Doe p/98765432 e/johnd@example.com rm/05-73 dr/19/05/20 - 24/05/22 ng/3 rc/no` |
+| **Clear**  | `clear`                                                                                                                                                                                                |
+| **Delete** | `delete INDEX`<br> e.g., `delete 3`                                                                                                                                                                    |
+| **Edit**   | `edit INDEX [n/NAME] [p/PHONE] [e/EMAIL] [dr/DATE_RANGE] [ng/NUMBER_OF_GUESTS]​`<br> e.g.,`edit 2 e/jameslee@example.com dr/13/09/22 - 15/09/22 ng/5`                                                  |
```
Should there be a [rm/ROOM] here?

kangzongxian replied:
Indeed!

bryanhce commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -52,37 +57,44 @@ public void parse_allFieldsPresent_success() {
+                new AddCommand(expectedGuest));
+
+        // multiple rooms - last room accepted
+        assertParseSuccess(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_AMY + EMAIL_DESC_BOB
```
Should there be only one email field here instead of EMAIL_DESC_AMY and EMAIL_DESC_BOB?

kangzongxian replied:
Rectified!

bryanhce commented on src/test/java/seedu/guest/model/guest/RoomTest.java:
```diff
@@ -0,0 +1,40 @@
+        // invalid room
+        assertFalse(Room.isValidRoom("")); // empty string
+        assertFalse(Room.isValidRoom(" ")); // spaces only
+        assertFalse(Room.isValidRoom("-50-23")); // not allowed to start with a hyphen
```
Is it important to have a case where the room number is "06--09", where there is 2 hyphens?

kangzongxian replied:
Sounds good!

bryanhce reviewed (changes requested):
Nice work! Just some nits to fix.

jehousoh commented on src/main/java/seedu/guest/model/guest/Room.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+
+
```
Just wondering is there any reasons for there to have so many newlines here?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // missing name prefix
-        assertParseFailure(parser, VALID_NAME_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + DATE_RANGE_DESC_BOB
+        assertParseFailure(parser, VALID_NAME_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // missing phone prefix
-        assertParseFailure(parser, NAME_DESC_BOB + VALID_PHONE_BOB + EMAIL_DESC_BOB + DATE_RANGE_DESC_BOB
+        assertParseFailure(parser, NAME_DESC_BOB + VALID_PHONE_BOB + EMAIL_DESC_BOB
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // missing email prefix
-        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + VALID_EMAIL_BOB + DATE_RANGE_DESC_BOB
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + VALID_EMAIL_BOB
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
 
+        // missing room prefix
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                        + VALID_ROOM_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // missing date range prefix
-        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + VALID_DATE_RANGE_BOB
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                        + ROOM_DESC_BOB + VALID_DATE_RANGE_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // missing number of guests prefix
-        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + DATE_RANGE_DESC_BOB
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // missing is room clean prefix
-        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + DATE_RANGE_DESC_BOB
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // all prefixes missing
-        assertParseFailure(parser, VALID_NAME_BOB + VALID_PHONE_BOB + VALID_EMAIL_BOB + VALID_DATE_RANGE_BOB
+        assertParseFailure(parser, VALID_NAME_BOB + VALID_PHONE_BOB + VALID_EMAIL_BOB
+                        + VALID_ROOM_BOB + VALID_DATE_RANGE_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // invalid name
-        assertParseFailure(parser, INVALID_NAME_DESC + PHONE_DESC_BOB + EMAIL_DESC_BOB + DATE_RANGE_DESC_BOB
+        assertParseFailure(parser, INVALID_NAME_DESC + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // invalid phone
-        assertParseFailure(parser, NAME_DESC_BOB + INVALID_PHONE_DESC + EMAIL_DESC_BOB + DATE_RANGE_DESC_BOB
+        assertParseFailure(parser, NAME_DESC_BOB + INVALID_PHONE_DESC + EMAIL_DESC_BOB
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // invalid email
-        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + INVALID_EMAIL_DESC + DATE_RANGE_DESC_BOB
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + INVALID_EMAIL_DESC
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
+        // invalid room
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                        + INVALID_ROOM_DESC + DATE_RANGE_DESC_BOB
+                        + NUMBER_OF_GUESTS_DESC_BOB + IS_ROOM_CLEAN_DESC_BOB,
```
Should the indentation of the line follow that of the assertParseSuccess above?

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // invalid date range
-        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + INVALID_DATE_RANGE_DESC
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                        + ROOM_DESC_BOB + INVALID_DATE_RANGE_DESC
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // invalid number of guests
-        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + DATE_RANGE_DESC_BOB
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // invalid is room clean
-        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB + DATE_RANGE_DESC_BOB
+        assertParseFailure(parser, NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // two invalid values, only first invalid value reported
-        assertParseFailure(parser, INVALID_NAME_DESC + PHONE_DESC_BOB + EMAIL_DESC_BOB + DATE_RANGE_DESC_BOB
+        assertParseFailure(parser, INVALID_NAME_DESC + PHONE_DESC_BOB + EMAIL_DESC_BOB
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/AddCommandParserTest.java:
```diff
@@ -91,81 +103,107 @@ public void parse_compulsoryFieldMissing_failure() {
         // non-empty preamble
         assertParseFailure(parser, PREAMBLE_NON_EMPTY + NAME_DESC_BOB + PHONE_DESC_BOB + EMAIL_DESC_BOB
-                + DATE_RANGE_DESC_BOB + NUMBER_OF_GUESTS_DESC_BOB + IS_ROOM_CLEAN_DESC_BOB,
+                        + ROOM_DESC_BOB + DATE_RANGE_DESC_BOB + NUMBER_OF_GUESTS_DESC_BOB + IS_ROOM_CLEAN_DESC_BOB,
```
Should the indentation of the line follow that of the assertParseSuccess above?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/logic/parser/EditCommandParserTest.java:
```diff
@@ -102,19 +109,21 @@ public void parse_invalidValue_failure() {
-        assertParseFailure(parser, "1" + INVALID_NAME_DESC + INVALID_EMAIL_DESC + VALID_DATE_RANGE_AMY
+        assertParseFailure(parser, "1" + INVALID_NAME_DESC + INVALID_EMAIL_DESC
+                + VALID_ROOM_AMY + VALID_DATE_RANGE_AMY
                         + VALID_PHONE_AMY, Name.MESSAGE_CONSTRAINTS);
```
Should the indentation of this line be the same as the line above it?

kangzongxian replied:
Rectified!

jehousoh commented on src/test/java/seedu/guest/model/guest/RoomTest.java:
```diff
@@ -0,0 +1,40 @@
+        assertFalse(Room.isValidRoom("-50-23")); // not allowed to start with a hyphen
+        assertFalse(Room.isValidRoom("$50")); // '$' is not allowed
+        assertFalse(Room.isValidRoom("^&*")); // special symbols only are not allowed
+        assertFalse(Room.isValidRoom("aloha-5()")); // '(' or ')' is not allowed
```
Should we also add the case where there is a space within the room number? eg 05 - 23

kangzongxian replied:
Sounds good!

jehousoh commented on src/test/java/seedu/guest/model/guest/RoomTest.java:
```diff
@@ -0,0 +1,40 @@
+
+        // valid room
+        assertTrue(Room.isValidRoom("05-73")); // only alphanumeric characters and hyphen
+        assertTrue(Room.isValidRoom("Aloha-5")); // only alphanumeric characters and hyphen
```
Should we also add the cases where there are no hyphens? ie All alphabets or all numbers

kangzongxian replied:
Sounds good!

jehousoh commented on src/test/java/seedu/guest/model/util/SampleDataUtilTest.java:
```diff
@@ -12,6 +12,8 @@
 import seedu.guest.model.guest.NumberOfGuests;
 import seedu.guest.model.guest.Phone;
+import seedu.guest.model.guest.Room;
+
```
Is this newline is necessary?

kangzongxian replied:
Rectified!

jehousoh reviewed (changes requested):
Looks good bar some styling and testing issues!

kangzongxian committed:
Fix code styling issues
Add more test cases for better coverage

kangzongxian committed:
Settle merge conflicts with master

Merge branch 'master' into add-room

* master:
  Increase flexibility of bill command input and add tests for negative total bill
  Enable assertion
  Add test classes for BillComand and BillCommandParser
  Update user guide with bill command
  Add bill command
  Add additional test cases for bill and validation message
  Fix wildcard imports
  Add BillTest class, modify Bill validation regex
  Add bill field to Guest

# Conflicts:
#	src/main/java/seedu/guest/logic/commands/EditCommand.java
#	src/main/java/seedu/guest/logic/parser/AddCommandParser.java
#	src/main/java/seedu/guest/logic/parser/CliSyntax.java
#	src/main/java/seedu/guest/logic/parser/ParserUtil.java
#	src/main/java/seedu/guest/model/guest/Guest.java
#	src/main/java/seedu/guest/model/util/SampleDataUtil.java
#	src/main/java/seedu/guest/storage/JsonAdaptedGuest.java
#	src/test/data/JsonSerializableGuestBookTest/typicalGuestGuestBook.json
#	src/test/java/seedu/guest/model/util/SampleDataUtilTest.java
#	src/test/java/seedu/guest/storage/JsonAdaptedGuestTest.java
#	src/test/java/seedu/guest/testutil/GuestBuilder.java
#	src/test/java/seedu/guest/testutil/TypicalGuests.java

bryanhce commented on src/main/java/seedu/guest/model/guest/Room.java:
```diff
@@ -0,0 +1,56 @@
+            + "E.g 05-73 or Aloha-5 are accepted.";
+
+    // Room contains alphanumeric characters with an optional hyphen only in between, and don't start with a hyphen
+    private static final String VALIDATION_REGEX = "^[A-Za-z0-9]+(?:-?[A-Za-z0-9]+)+$";
```
Can I check if this VALIDATION_REGEX accepts "09-09-09"? If yes, the MESSAGE_CONSTRAINTS description might have missed that part out?

kangzongxian replied:
Indeed, added!

bryanhce reviewed (commented):
Nice!

kangzongxian committed:
Update comments regarding Room

bryanhce reviewed (approved):
LGTM

jehousoh reviewed (approved):
lgtm
